src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonObjectDefaultConverter.cs

```diff
@@ -0,0 +1,410 @@
+                    {
+                        ThrowHelper.ThrowNotSupportedException_DeserializeNoParameterlessConstructor(state.Current.JsonClassInfo.Type);
+                    }
+
+                    obj = state.Current.JsonClassInfo.CreateObject!()!;
+                    if (state.Current.MetadataId != null)
+                    {
+                        if (!state.ReferenceResolver.AddReferenceOnDeserialize(state.Current.MetadataId, obj))
+                        {
+                            ThrowHelper.ThrowJsonException_MetadataDuplicateIdFound(state.Current.MetadataId, ref state);
+                        }
+                    }
+
+                    state.Current.ReturnValue = obj;
+                    state.Current.ObjectState = StackFrameObjectState.CreatedObject;
+                }
+                else
+                {
+                    obj = state.Current.ReturnValue!;
+                    Debug.Assert(obj != null);
+                }
+
+                // Read all properties.
+                while (true)
+                {
+                    // Determine the property.
+                    if (state.Current.PropertyState == StackFramePropertyState.None)
+                    {
+                        state.Current.PropertyState = StackFramePropertyState.ReadName;
+
+                        if (!reader.Read())
+                        {
+                            // The read-ahead functionality will do the Read().
+                            state.Current.ReturnValue = obj;
+                            value = default!;
+                            return false;
+                        }
+                    }
+
+                    JsonPropertyInfo jsonPropertyInfo;
+
+                    if (state.Current.PropertyState < StackFramePropertyState.Name)
+                    {
+                        state.Current.PropertyState = StackFramePropertyState.Name;
+
+                        JsonTokenType tokenType = reader.TokenType;
+                        if (tokenType == JsonTokenType.EndObject)
+                        {
+                            // We are done reading properties.
+                            break;
+                        }
+                        else if (tokenType != JsonTokenType.PropertyName)
+                        {
+                            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                        }
+
+                        jsonPropertyInfo = JsonSerializer.LookupProperty(
+                            obj,
+                            ref reader,
+                            options,
+                            ref state,
+                            out bool useExtensionProperty);
+
+                        state.Current.UseExtensionProperty = useExtensionProperty;
+                    }
+                    else
+                    {
+                        jsonPropertyInfo = state.Current.JsonPropertyInfo!;
```

Comment:
What's guaranteeing this is not null?

Reply:
The LookupProperty method was called previously which sets `state.Current.JsonPropertyInfo` to a non-null value. I can add an assert\comments.